CppExport/src/exporter/CppExporter.cpp

```diff
@@ -34,74 +34,157 @@
+{
+	if (auto ooModule = DCast<OOModel::Module>(current))
 	{
-		auto c = project->classes()->at(i);
-		//if (c->name() == "HelloWorld" || c->name() == "Generic")
+		if (ooModule->classes()->size() > 0)
+			namespaceName = ooModule->name();
+		else
 		{
-			auto unit = new CodeUnit(c->name(), c);
-			codeComposite->addUnit(unit);
+			// macro file
+			// TODO: handle non class units
+			//result.append(new CodeUnit((namespaceName.isEmpty() ? "" : namespaceName + "/") + ooModule->name(), current));
+			return;
 		}
 	}
+	else if (auto ooClass = DCast<OOModel::Class>(current))
+	{
+		result.append(new CodeUnit((namespaceName.isEmpty() ? "" : namespaceName + "/") + ooClass->name(), current));
+		return;
+	}
 
-	exportCodeComposite(codeComposite);
-
-	return {};
+	for (auto child : current->children())
+		units(child, namespaceName, result);
+}
 
-	auto layout = layouter();
+QList<CodeComposite*> CppExporter::mergeUnits(QList<CodeUnit*>& units)
+{
+	QHash<QString, QString> mergeMap = Config::instance().dependencyUnitMergeMap();
 
-	DeclarationVisitorHeader headerVisitor;
-	auto dir = std::unique_ptr<Export::SourceDir>( headerVisitor.visitProject(project) );
-	auto map = Export::Exporter::exportToFileSystem(pathToProjectContainerDirectory + "/Headers", dir.get(), &layout);
-	exportMaps().insert(project, map);
+	QHash<QString, CodeComposite*> nameToCompositeMap;
+	for (auto unit : units)
+	{
+		auto it = mergeMap.find(unit->name());
+		auto compositeName = it != mergeMap.end() ? *it : unit->name();
 
-	DeclarationVisitorSource sourceVisitor;
-	auto dir2 = std::unique_ptr<Export::SourceDir>( sourceVisitor.visitProject(project) );
-	auto map2 = Export::Exporter::exportToFileSystem(pathToProjectContainerDirectory + "/Sources", dir2.get(), &layout);
-	exportMaps().insert(project, map2);
+		auto cIt = nameToCompositeMap.find(compositeName);
+		if (cIt != nameToCompositeMap.end())
+			// case A: the composite that unit is a part of already exists => merge
+			(*cIt)->addUnit(unit);
+		else
+		{
+			// case B: the composite that unit is a part of does not yet exist
+			auto composite = new CodeComposite(compositeName);
+			composite->addUnit(unit);
+			nameToCompositeMap.insert(composite->name(), composite);
+		}
+	}
 
-	return headerVisitor.errors();
+	return nameToCompositeMap.values();
 }
 
-void CppExporter::exportFragment(Export::SourceFragment* fragment)
+void CppExporter::calculateSourceFragments(QList<CodeComposite*> codeComposites)
 {
-	auto dir = new Export::SourceDir(nullptr, "src");
-	auto file = &dir->file("exported.h");
-	file->append(fragment);
-	auto layout = layouter();
-	Export::Exporter::exportToFileSystem("", dir, &layout);
+	DeclarationVisitorHeader headerVisitor;
+	DeclarationVisitorSource sourceVisitor;
+
+	for (auto codeComposite : codeComposites)
+	{
+		for (auto unit : codeComposite->units())
+			if (auto classs = DCast<OOModel::Class>(unit->node()))
```

Comment:
Also the loop in `calculateSourceFragments` should be nothing more than a call to a `calculateSourceFragments` method inside each Unit which contains this if statement.